lib/Convos/Core/Backend/File.pm

```diff
@@ -80,14 +80,26 @@ sub messages {
   $re = qr{\Q$re\E}i unless ref $re;
   $re = qr/^(\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}) (.*$re.*)$/;
 
-  $args{after}  = Time::Piece->strptime($query->{after},  '%Y-%m-%dT%H:%M:%S') if $query->{after};
-  $args{before} = Time::Piece->strptime($query->{before}, '%Y-%m-%dT%H:%M:%S') if $query->{before};
-
-  # If neither before nor after are defined, search everything before now
-  $args{before} = gmtime if !$args{before} && !$args{after};
   $args{limit}    = $query->{limit} || 60;
   $args{messages} = [];
   $args{re}       = $re;
+  if ( !$query->{before} && !$query->{after} ) {
+    $args{before} = gmtime;
+    $args{after} = $args{before}->add_months(-12);
+  } elsif ( $query->{before} && !$query->{after} ) {
+    $args{before} = Time::Piece->strptime($query->{before}, '%Y-%m-%dT%H:%M:%S');
+    $args{after} = $args{before}->add_months(-12);
+  } elsif ( !$query->{before} && $query->{after} ) {
+    $args{after} = Time::Piece->strptime($query->{after}, '%Y-%m-%dT%H:%M:%S');
+    $args{before} = $args{after}->add_months(12);
+  } elsif ( $query->{before} && $query->{after} ) {
+    $args{before} = Time::Piece->strptime($query->{before}, '%Y-%m-%dT%H:%M:%S');
+    $args{after} = Time::Piece->strptime($query->{after}, '%Y-%m-%dT%H:%M:%S');
+    # return no messages if args > before or before - args > 12
+    $args{after} = $args{before}->add_months(1) if $args{before} - $args{after} > $args{before} - $args{before}->add_months(-12);
+  }
+  $args{cursor} = $args{before};
```

Comment:
I think the if/else/elsif above could collapse to:

```
$args{before}
  = $query->{before} ? Time::Piece->strptime($query->{before}, '%Y-%m-%dT%H:%M:%S') : gmtime;

$args{after}
  = $query->{after}
  ? Time::Piece->strptime($query->{after}, '%Y-%m-%dT%H:%M:%S')
  : $args{before}->add_months(-12);
```

It's fine that if someone just specifies "after" that they can search from "after" until "now". If we want to limit it further, we can do so in the JavaScript.

I don't get the last "return no messages if args > before or before - args > 12" statement.

The only reason for my change is that if it's easier to understand the code, it's easier to not mess up. But I'm not going to block this because you got a lot of tests, which makes it safe either way.